src/p4est_io.h

```diff
@@ -68,6 +108,398 @@ p4est_t            *p4est_inflate (sc_MPI_Comm mpicomm,
+ * This function leaves the file open if MPI I/O is available.
+ * It is necessary to call \ref
+ * p4est_file_close (possibly after writing one or more data sets).
+ * The file is opened in a write-only mode.
+ *
+ * We add some basic metadata to the file.
+ * The file written contains the file header and data sets
+ * as specified by the open/write functions called.
+ * The file header consists of the metadata specified by p4est.
+ *
+ * It is the application's responsibility to write sufficient header
+ * information (cf. \ref p4est_file_write_header) to determine the number and
+ * size of the data sets if such information is not recorded and maintained
+ * externally.
+ * However, p4est makes some metadata accessible via
+ * \ref p4est_file_info.
+ *
+ * This function does not abort on MPI I/O errors but returns NULL.
+ *
+ * \param [in] p4est          Valid forest.
+ * \param [in] filename       Path to parallel file that is to be created.
+ * \param [in] user_string    A user string that is written to the file header.
+ *                            Only 15 bytes without null-termination are
+ *                            written to the file. If the user gives less
+ *                            bytes the user_string in the file header is padded
+ *                            by spaces.
+ * \param [out] errcode       An errcode that can be interpreted by \ref
+ *                            p4est_file_error_string and
+ *                            \ref p4est_file_error_class.
+ * \return                    Newly allocated context to continue writing
+ *                            and eventually closing the file. NULL in
+ *                            case of error.
+ */
+p4est_file_context_t *p4est_file_open_create
+  (p4est_t * p4est, const char *filename, const char user_string[15],
+   int *errcode);
+
+/** Open a file for reading and read its user string on rank zero.
+ * The user string is broadcasted to all ranks after reading.
+ * The file must exist and be at least of the size of the file header.
+ *
+ * If the file has wrong metadata the function reports the error using
+ * /ref P4EST_LERRORF, collectively close the file and deallocate
+ * the file context. In this case the function returns NULL on all ranks.
+ * The wrong file format or a wrong file header causes \ref P4EST_ERR_IO
+ * as errcode.
+ *
+ * This function does not abort on MPI I/O errors but returns NULL.
+ *
+ * \param [in] p4est            The forest must be of the same refinement
+ *                              pattern as the one used for writing the file.
+ *                              Its global number of quadrants must match.
+ *                              It is possible, however, to use a different
+ *                              partition or number of ranks from writing it.
+ * \param [in] filename         The path to the file that is opened.
+ * \param [in,out] user_string  At least 16 bytes. The user string is written
```

Comment:
If we make this a `#define`, then we can use `\ref` in the documentation. Again, I'd like to decide generally to use `[NUMBER] `or `char *`, but not a mix of the two if that can be avoided.

Reply:
On both input and output, for strings, allow for the space of the terminating `NUL`, demand it on input and always write it on output.

Reply:
> On both input and output, for strings, allow for the space of the terminating `NUL`, demand it on input and always write it on output.

This is now done for input strings as described above and was already the case for the output strings.